### PR TITLE
[Model] Add LingBot-Vision Model

### PR DESCRIPTION
**Pull request by IMvision12**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47365)

Add LingBot-Vision

  What does this PR do?

  Adds LingBot-Vision, a self-supervised ViT vision backbone for dense spatial perception, trained with masked boundary modeling on top of a DINOv3-style architecture (RoPE,
  storage tokens, SwiGLU, LayerScale).

  - Paper: Vision Pretraining for Dense Spatial Perception (https://arxiv.org/abs/2607.05247) (arXiv 2607.05247)
  - Original code/weights: https://github.com/robbyant/lingbot-vision

colab gist : https://colab.research.google.com/drive/1yMQodhjpxN7CuFJln8LqsFbPDaJgLsEb?usp=sharing
  
lingbot_pca_grid

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @yonigozlan @molbap

File: src/transformers/models/lingbot_vision/modular_lingbot_vision.py [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+            coords_w = torch.arange(0.5, width, **kwargs) / max_hw
+        elif self.normalize_coords == "min":
+            min_hw = min(height, width)
+            coords_h = torch.arange(0.5, height, **kwargs) / min_hw
+            coords_w = torch.arange(0.5, width, **kwargs) / min_hw
+        elif self.normalize_coords == "separate":
+            coords_h = torch.arange(0.5, height, **kwargs) / height
+            coords_w = torch.arange(0.5, width, **kwargs) / width
+        else:
+            raise ValueError(f"Unknown RoPE coordinate normalization: {self.normalize_coords}")
```
Comment (molbap) [inline src/transformers/models/lingbot_vision/modular_lingbot_vision.py]: Rope parameterization is handled via `rope_parameters`

File: src/transformers/models/lingbot_vision/modular_lingbot_vision.py [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+                if self.config.apply_layernorm:
+                    if self.config.untie_cls_and_patch_norms:
+                        hidden_state = torch.cat(
+                            (
+                                self.cls_norm(hidden_state[:, :prefix_length]),
+                                self.layernorm(hidden_state[:, prefix_length:]),
+                            ),
+                            dim=1,
+                        )
+                    else:
```
Comment (molbap) [inline src/transformers/models/lingbot_vision/modular_lingbot_vision.py]: let's combine conditions for readability a bit

File: src/transformers/models/lingbot_vision/convert_lingbot_vision_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    config = get_lingbot_vision_config(variant)
+    model = LingbotVisionModel(config)
+    model.eval()
+
+    state_dict = load_original_state_dict(checkpoint_path, variant, cache_dir=cache_dir, revision=revision)
+
+    for src, dest in create_rename_keys(config):
+        if src in state_dict:
+            rename_key(state_dict, src, dest)
+
```
Comment (molbap) [inline src/transformers/models/lingbot_vision/convert_lingbot_vision_to_hf.py]: since this is mostly renaming (and even in case of more complicated ops) let's use `conversion_mapping` and the `WeightConverter` ops to have an on-the-fly conversion

File: tests/models/lingbot_vision/test_modeling_lingbot_vision.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        expected_seq_length = num_patches + 1 + model.config.num_storage_tokens
+        expected_shape = torch.Size((1, expected_seq_length, model.config.hidden_size))
+        self.assertEqual(outputs.last_hidden_state.shape, expected_shape)
+
+        expected_pooler = torch.tensor([-0.3808, 0.8251, 0.3735, 1.1332, 1.7006], device=torch_device)
+        torch.testing.assert_close(outputs.pooler_output[0, :5], expected_pooler, rtol=1e-4, atol=1e-4)
+
+        first_patch_token = outputs.last_hidden_state[:, model.config.num_storage_tokens + 1 :]
+        expected_patch = torch.tensor([0.5824, 0.2652, -0.3335, -0.0979, 0.2420], device=torch_device)
+        torch.testing.assert_close(first_patch_token[0, 0, :5], expected_patch, rtol=1e-4, atol=1e-4)
```
Comment (molbap) [inline tests/models/lingbot_vision/test_modeling_lingbot_vision.py:256]: nice, but let's use the `Expectations` class

File: src/transformers/models/lingbot_vision/modular_lingbot_vision.py [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+            context_layer = torch.matmul(attention_probs, value)
+        else:
+            attention_probs = None
+            context_layer = F.scaled_dot_product_attention(
+                query,
+                key,
+                value,
+                dropout_p=self.attention_dropout if self.training else 0.0,
+                scale=self.scale,
+            )
```
Comment (molbap) [inline src/transformers/models/lingbot_vision/modular_lingbot_vision.py]: we have an `attention_interface` called with `attention_interface: Callable = ALL_ATTENTION_FUNCTIONS.get_interface(` and the output of attentions/tracing is handled through decorators. see `model_output_tracing.md` documentation

**Review by molbap — COMMENTED** [review]
Hi @IMvision12 , thanks for the PR! I started reviewing, several patterns are not up-to-date/not using fully the current API. If you are using a coding agent I suggest using this experimental [self-review skill ](https://huggingface.co/datasets/Molbap/Skills/blob/main/transformers_contribution_skill.md) to do a run first, and then ping us back. We're trying out this in order to facilitate review, merge and maintenance of new models and lib components 🤗

File: src/transformers/models/lingbot_vision/modular_lingbot_vision.py (hunk 1 above)
Comment (IMvision12) [reply src/transformers/models/lingbot_vision/modular_lingbot_vision.py]: Done

**Comment by IMvision12** [comment]
Updated colab gist : https://colab.research.google.com/drive/1yMQodhjpxN7CuFJln8LqsFbPDaJgLsEb?usp=sharing

**Comment by molbap** [comment]
Hello @IMvision12 , have you run the self-review skill? https://huggingface.co/datasets/Molbap/Skills/blob/main/transformers_contribution_skill.md ? It looks like the current model has DINOv3-style architecture (RoPE, storage tokens, SwiGLU, LayerScale)," dinov3_vit exists with an identical structure, yet the modular imports Llama/Pixtral/Swin and reimplements ~10 classes that have direct DINOv3ViT counterparts, can you take another look and answer? Thank you.

**Comment by IMvision12** [comment]
@molbap Thank you, I initially implemented the blocks locally by hand and then ran the self-review skill. I missed the extent of the DINOv3ViT overlap, so I’ll revisit the modular implementation and reuse its existing components wherever possible.

After the rewrite, modular_lingbot_vision.py is 131 lines and the modeling file is DINOv3ViT with names changed  I verified an unmodified DINOv3ViTModel runs the released ViT-S/16 weights at 3.8e-6 from the reference. So the honest position is that LingBot-Vision is DINOv3ViT, retrained with masked boundary modeling.

So is it worth adding new model type or  publish checkpoints as dinov3_vit with a convert script

**Comment by molbap** [comment]
Hi @IMvision12 , given
> an unmodified DINOv3ViTModel runs the released ViT-S/16 weights at 3.8e-6 from the reference.

I think if all checkpoints are the same, then the best course of action is adding the correct tags (transformers, dinov3_vit) to the original organisation's repos: https://huggingface.co/robbyant e.g. https://huggingface.co/robbyant/lingbot-vision-vit-base , it seems there's an ongoing discussion there. https://huggingface.co/robbyant/lingbot-vision-vit-giant/discussions/2

**Comment by IMvision12** [comment]
Closing this then .
